Use Dtree for distributed parallelism

kpamnany opened the pull request:
This will run as before on a single node and additionally output timing. Multi-node support is in progress.

Done:
- Using Dtree to distribute work.
- Add load imbalance measurement.
- Added timing for various Celeste parts.

TODO:
- Tune work item ra/dec based on load imbalance, other criteria?
- Use a thread to drive Dtree in parent nodes.

jeff-regier commented:
Bugs #168 and #169 are resolved. Ryan wanted a unit test for them, but that's not a priority imo given the timeline we're on, and because I've tested the new code pretty thoroughly. Bugs #175 and #177 persist, but #175 doesn't look like a bad one at least. I'm trying to further pin down #177.

kpamnany commented:
Stuck on a Julia multithreading bug (https://github.com/JuliaLang/julia/issues/15620). This won't interfere with multi-node scaling, on which I am continuing work.

kpamnany commented:
This isn't completely ready yet. I still have to add some instrumentation to report load imbalance and clean up the timing output so it isn't so messy. Hopefully by EOD, then the output can be verified for correctness as follows.

Grab a Cori node with `salloc -N 1 -p debug -t 00:30:00`. Make sure to `export JULIA_NUM_THREADS=1`. Then you can run multi-process with `mpiexec.hydra -np X bin/celeste.jl infer-nersc 200.0 200.05 38.1 38.15 $SCRATCH` where `X` is the number of processes you want (up to 32 on a single node).

True multi-node runs should also work, but I haven't tested those yet. Right now, for the 27 sources in this patch of sky, it looks like 659 seconds with 1 process, 309 seconds with 2 processes (super-linear? how?), and 201 seconds with 4 processes, all with static scheduling.

kpamnany commented:
Okay. I think this is ready. I'll work on multi-threading and other improvements on another branch.

Commit:
Use Dtree for distributed parallelism

Single node execution shouldn't change except for a lot more timing
information displayed. Need to:
Pkg.clone("https://github.com/kpamnany/Dtree.jl")
Pkg.build("Dtree")

Not currently using threads, so load balance isn't great yet because
parent nodes aren't responding to child requests quickly. Converted
some println()s to info()s. Fixed reduce_elbo_vars_array!().

kpamnany commented:
Note: there are a number of ugly hacks here. There's probably a better place for `nputs`. Passing `nid` and `nnodes` around is clunky. As @rgiordan has pointed out, the `Threaded` global in `ElboDeriv.jl` should ideally not be a global there. I'm not sure how best to resolve some of these, so feel free to fix/suggest fixes.

rgiordan commented on bin/celeste.jl:
```diff
@@ -1,8 +1,14 @@
 
+const cpu_hz = 2.3e9
+
+const wira = 0.015
```
These are the width of ra / dec rectangles.  Could you please add a comment to this effect?

rgiordan commented on src/ModelInit.jl:
```diff
@@ -671,7 +671,7 @@ function trim_source_tiles(
       bright_pixels = Bool[];
       if has_source
-        #println("Tiles $h $w has source $s")
+        info("Tiles $h $w has source $s")
```
I don't think you want to display this -- it was just for debugging.  My bad for commenting it out and not just deleting it.

rgiordan commented on src/Transform.jl:
```diff
@@ -890,7 +890,7 @@ function enforce_bounds!{NumType <: Number}(
              mp.vp[s][ids.(param)[ind]] <=
              constraint.upper_bound)
-          println("Warning: $param[$s][$ind] was out of bounds.")
+          info("Warning: $param[$s][$ind] was out of bounds.")
```
info("Warning:...") seems like a strange thing to say.

kpamnany replied:
I was just removing all the println()s. I'll let you guys decide if these info/warning/print messages are required.

rgiordan commented on src/api.jl:
```diff
@@ -161,13 +182,14 @@ function infer(fieldids::Vector{Tuple{Int, Int, Int}},
     # If there are no objects of interest, return early.
     if length(target_sources) == 0
-        return Dict{Int, Dict}()
+        return Dict{Int, Dict}(), (0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0)
```
It would be nice to put all these times in a data structure.

kpamnany replied:
Where should I define this type (and other stuff), so that it is visible in `bin/celeste.jl` and in `src/api.jl`?

rgiordan replied:
I don't think it needs to be visible in `bin/celeste.jl`, does it?  In either case, simply defining it in `src/api.jl` should do since (for now at least) it's simply included in the Celeste module.

kpamnany replied:
Some additional time measurements are made in `bin/celeste.jl` and they are all output there. Also, I want to put `nputs` and other such stuff that is used in both places somewhere... putting stuff in `src/api.jl` makes it necessary to add the `Celeste.` prefix, or to explicitly export things. :(

rgiordan replied:
Well, IMO you're putting things in `bin/celeste.jl` that should be in `api.jl` (and tested), and there are things in `api.jl` that should be in `SDSSIO.jl` (or something like it).  Again, IMO, timing measures are attributes of an API and should be defined with the API in its own module, not in a binary script.

After the paper I think we should organize the whole thing better.  In the meantime, just go ahead and do what you think is best that gets you up and running the fastest that doesn't interfere with the rest of us by moving stuff around too much, and we'll be fine.  :)

jeff-regier replied:
It's fine to hold off on writing unit tests, but I do think it's going to be confusing if `celeste.jl` contains code that isn't just about parsing arguments and calling the right function in `api.jl`.

rgiordan commented on src/api.jl:
```diff
@@ -15,6 +30,7 @@ const TILE_WIDTH = 20
 const DEFAULT_MAX_ITERS = 50
 const MIN_FLUX = 2.0
 
+@inline nputs(nid, s) = ccall(:puts, Cint, (Ptr{Int8},), string("[$nid] ", s))
```
Just curious, why use this and not Logging?

kpamnany replied:
This runs "atomically", so the text from multiple nodes isn't interleaved. Also, Logging is used extensively all over Celeste, and I need to turn all that off otherwise there is too much screen spam.

rgiordan commented on src/ElboDeriv.jl:
```diff
@@ -11,8 +11,8 @@ import ..KL
 
-Threaded = true
-if VERSION > v"0.5.0-dev"
+Threaded = false
```
Is the threaded stuff in ElboDeriv.jl used at all anymore?  If not, are you just leaving it in in case we change our mind in the future?  If so, could you maybe mention that in a comment?

rgiordan commented on bin/celeste.jl:
```diff
@@ -27,7 +33,175 @@ field, using only that field.
+
+    n_infers = 0
+    n_fits = 0
+    tm_infer = tm_fldids = tm_rdprim = tm_filcat = tm_rdimg = tm_imp = tm_psf = tm_fit = tm_out = tm_done = 0.0
```
There are a lot of line overruns and tab misalignments -- I won't comment each one.

kpamnany replied:
I think I fixed all of them.

rgiordan commented on bin/celeste.jl:
```diff
@@ -27,7 +33,175 @@ field, using only that field.
+
+This function assumes a cartesian (rather than spherical) coordinate system!
+"""
+function sky_subarea(skya::SkyArea, i)
```
Just my opinion, but I think it's good style to give functions verb names and variables noun names.

rgiordan commented on bin/celeste.jl:
```diff
@@ -27,7 +33,175 @@ field, using only that field.
+"""
+function celeste_skyarea(ramin, ramax, decmin, decmax, outdir)
+    # ---
+    if dt_nodeid == 1
```
Where are `dt_nodeid` and `dt_nnodes` set?

kpamnany replied:
They come in from Dtree.

rgiordan commented:
This looks good to me substantively (without knowing the workings of Dtree).  I just had some superficial comments and questions -- we can always tidy up after the deadline.  :)

Commit:
Minor mostly cosmetic cleanups

Commit:
Testing multithreaded pixel loop

Commit:
Multithread the sources loop.

kpamnany commented:
Closing this; will open a new PR with a refactored implementation.
